azurerm/resource_arm_nat_gateway.go

```diff
@@ -0,0 +1,306 @@
+package azurerm
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strconv"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-06-01/network"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/response"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/validate"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/features"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tags"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+var natGatewayResourceName = "azurerm_nat_gateway"
+
+func resourceArmNatGateway() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmNatGatewayCreateUpdate,
+		Read:   resourceArmNatGatewayRead,
+		Update: resourceArmNatGatewayCreateUpdate,
+		Delete: resourceArmNatGatewayDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.NoEmptyStrings,
```

Comment:
Can we add some real validation here other than `NoEmptyStrings`?

Reply:
This field is `name`, no stronger constraint than `NoEmpty` I can think of. Do you have a suggestion about this?